### PR TITLE
LOK-1863 Event Service: Add Service level test cases 

### PR DESCRIPTION
**Pull request by jahanzebdataq**
## Description
vent Service is missing service level test CucumberIT

we should cover

Trap being sent → Event being persisted

Trap being sent for known node → Event having node info

Trap being sent for unknown node → New suspect event being sent

Receiving external events, internal events → Events being persisted

## Jira link(s)
- https://opennms.atlassian.net/browse/LOK-1863

### CODE REVIEW

File: events/docker-it/src/test/java/org/opennms/horizon/events/CucumberRunnerIT.java [hunk 1]
```diff
@@ -89,18 +105,18 @@ private static void startApplicationContainer(boolean enableDebuggingPort5005) {
...
+                        "SPRING_KAFKA_BOOTSTRAP_SERVERS",
+                        kafkaContainer.getNetworkAliases().get(0) + ":9092")
                 .withEnv("SPRING_DATASOURCE_URL", jdbcUrl)
                 .withEnv("SPRING_DATASOURCE_USERNAME", postgreSQLContainer.getUsername())
                 .withEnv("SPRING_DATASOURCE_PASSWORD", postgreSQLContainer.getPassword())
-                .withEnv("INVENTORY_ENCRYPTION_KEY", RandomStringUtils.randomAlphanumeric(32))
-                // Uncomment to get Hibernate SQL logging
-                // .withEnv("logging.level.org.hibernate.SQL", "DEBUG")
-                // .withEnv("logging.level.org.hibernate.orm.jdbc.bind", "TRACE")
+                .withEnv("EVENT_ENCRYPTION_KEY", RandomStringUtils.randomAlphanumeric(32))
```
Comment (cgorantla) [inline events/docker-it/src/test/java/org/opennms/horizon/events/CucumberRunnerIT.java:119]: don't need this env `"EVENT_ENCRYPTION_KEY`

File: events/docker-it/pom.xml [hunk 2]
```diff
@@ -182,6 +182,11 @@
       <artifactId>kafka-clients</artifactId>
       <version>3.4.0</version>
     </dependency>
+    <dependency>
```
Comment (cgorantla) [inline events/docker-it/pom.xml:185]: where are you using this ?

File: events/docker-it/src/test/java/org/opennms/horizon/events/stepdefs/EventStepDefinitions.java [hunk 3]
```diff
@@ -62,4 +74,59 @@ public void verifyThereAreEvents(int arg0) {
...
+                .pollDelay(1, TimeUnit.SECONDS)
+                .pollInterval(2, TimeUnit.SECONDS)
+                .until(
+                        () ->
+                                backgroundHelper
+                                        .getEventServiceBlockingStub()
+                                        .searchEvents(searchEventByLocationName)
+                                        .getEventsList()
+                                        .stream()
+                                        .anyMatch(event -> event.getTenantId().equals(this.tenantId)),
```
Comment (cgorantla) [inline events/docker-it/src/test/java/org/opennms/horizon/events/stepdefs/EventStepDefinitions.java:117]: It's better to check if it produces specific event with uei, not just tenantId

File: events/docker-it/src/test/resources/org/opennms/horizon/events/event.feature [hunk 4]
```diff
@@ -6,5 +6,7 @@ Feature: Event Service Basic Functionality
     Given [Event] Grpc TenantId "event-tenant-stream"
     Given [Event] Create Grpc Connection for Events
 
-    Scenario: Get event numbers
-      Then verify there are 0 events
+  Scenario:
+    Given Initialize Trap Producer With Topic "traps" and BootstrapServer "kafka.bootstrap-servers"
+    When Send Trap Data to Kafka Listener via Producer with TenantId "event-tenant-stream" and Location "1"
```
Comment (cgorantla) [inline events/docker-it/src/test/resources/org/opennms/horizon/events/event.feature:11]: Location with some name will be meaningful

Comment (jahanzebdataq) [reply events/docker-it/src/test/resources/org/opennms/horizon/events/event.feature:11]: If we give name here, then we have to find out location id through location name, for which we have to use MonitoringLocationServiceGrpc from inventory service

File: events/docker-it/src/test/java/org/opennms/horizon/events/stepdefs/EventStepDefinitions.java (hunk 3 above)
Comment (jahanzebdataq) [reply events/docker-it/src/test/java/org/opennms/horizon/events/stepdefs/EventStepDefinitions.java:117]: What is uei? and how do we get and store it in our class to be checked against it?

File: events/docker-it/src/test/java/org/opennms/horizon/events/CucumberRunnerIT.java (hunk 1 above)
Comment (jahanzebdataq) [reply events/docker-it/src/test/java/org/opennms/horizon/events/CucumberRunnerIT.java:119]: Resolved

File: events/docker-it/pom.xml (hunk 2 above)
Comment (jahanzebdataq) [reply events/docker-it/pom.xml:185]: Removed this dependency

File: events/docker-it/src/test/java/org/opennms/horizon/events/stepdefs/EventStepDefinitions.java (hunk 3 above)
Comment (jahanzebdataq) [reply events/docker-it/src/test/java/org/opennms/horizon/events/stepdefs/EventStepDefinitions.java:117]: Added a uei check

File: events/docker-it/src/test/java/org/opennms/horizon/events/stepdefs/EventStepDefinitions.java [hunk 5]
```diff
@@ -62,4 +75,61 @@ public void verifyThereAreEvents(int arg0) {
...
+
+    @When("Send Trap Data to Kafka Listener via Producer with TenantId {string} and Location {string}")
+    public void sendTrapDataToKafkaListenerViaProducerWithTenantIdAndLocationId(String tenantId, String location) {
+        TenantLocationSpecificTrapLogDTO tenantLocationSpecificTrapLogDTO =
+                TenantLocationSpecificTrapLogDTO.newBuilder()
+                        .setTenantId(tenantId)
+                        .setLocationId(location)
+                        .setTrapAddress("127.0.0.1")
+                        .addTrapDTO(TrapDTO.newBuilder()
+                                .setAgentAddress("127.0.0.1")
```
Comment (cgorantla) [inline events/docker-it/src/test/java/org/opennms/horizon/events/stepdefs/EventStepDefinitions.java:87]: Not sure how does it produce an event with uei `uei.opennms.org/generic/traps/SNMP_Cold_Start` when you don't have any relevant Trap in TrapDTO, Can you re-check this ?

**Comment by cgorantla** [comment]
This is consistently failing at my end.

**Comment by jahanzebdataq** [comment]
> This is consistently failing at my end.

Can you show me the error so we can see why it is failing at your end, even though it ran perfectly here

**Review by cgorantla — COMMENTED** [review]
This doesn't seems to work.
None of the trap are received on Event Service.

Please debug using remote debug by enabling debug port

**Comment by jahanzebdataq** [comment]
> This doesn't seems to work. None of the trap are received on Event Service.
> 
> Please debug using remote debug by enabling debug port

I have tested it on my local machine, also changed the parameters of trap being sent to trap listener, after getting back event list, we get the event with same parameters. Please view the video. Also I committed these changes and all check in the PR were passed and it executed related cucumber tests.
[Screencast Event Service Trap to event test.webm](https://github.com/OpenNMS-Cloud/lokahi/assets/159922670/de6fe5e9-339b-4606-9550-a53748568edf)

**Comment by jahanzebdataq** [comment]
> This doesn't seems to work. None of the trap are received on Event Service.
> 
> Please debug using remote debug by enabling debug port

I was not able to debug and find out whether trap was being sent and persisted or not, I added the logs in event consumer as you can see in the images below, upon executing the code you can see the logs of events being sent by trap consumer and persisted by event consumer.
Screenshot from 2024-05-07 22-58-14
Screenshot from 2024-05-07 23-02-57
Screenshot from 2024-05-07 23-08-07

File: events/consumer/src/main/java/org/opennms/horizon/events/consumer/EventsConsumer.java [hunk 6]
```diff
@@ -64,6 +64,7 @@ public void consume(@Payload byte[] data) {
         try {
             EventLog eventLog = EventLog.parseFrom(data);
             LOG.trace("Received events from kafka {}", eventLog);
+            LOG.info("Received events from kafka {}", eventLog);
```
Comment (fooker) [inline events/consumer/src/main/java/org/opennms/horizon/events/consumer/EventsConsumer.java:67]: Why both?

File: events/docker-it/src/test/java/org/opennms/horizon/events/CucumberRunnerIT.java [hunk 7]
```diff
@@ -67,14 +75,21 @@ public static void before() throws Throwable {
...
+        kafkaContainer.start();
         postgreSQLContainer.start();
+
+        String bootstrapServers = kafkaContainer.getBootstrapServers();
+        System.setProperty(KAFKA_BOOTSTRAP_SERVER_PROPERTYNAME, bootstrapServers);
+        LOG.info("KAFKA LOCALHOST BOOTSTRAP SERVERS {}", bootstrapServers);
+
         startApplicationContainer(
-                false); // DEBUGGING - set to true to expose the application debugging on host port 5005
+                true); // DEBUGGING - set to true to expose the application debugging on host port 5005
```
Comment (fooker) [inline events/docker-it/src/test/java/org/opennms/horizon/events/CucumberRunnerIT.java:86]: Please revert

**Review by fooker — CHANGES_REQUESTED** [review]

File: events/consumer/src/main/java/org/opennms/horizon/events/consumer/EventsConsumer.java (hunk 6 above)
Comment (jahanzebdataq) [reply events/consumer/src/main/java/org/opennms/horizon/events/consumer/EventsConsumer.java]: Removed

**Review by fooker — APPROVED** [review]

**Pull request merged**